hatchet_sdk/v2/hatchet.py

```diff
@@ -44,6 +57,56 @@ def transform_desired_worker_label(d: DesiredWorkerLabel) -> DesiredWorkerLabels
     )
 
 
+class Function(Generic[R, TWorkflowInput]):
+    def __init__(
+        self,
+        fn: Callable[[Context], R],
+        hatchet: "Hatchet",
+        name: str = "",
+        on_events: list[str] = [],
+        on_crons: list[str] = [],
+        version: str = "",
+        timeout: str = "60m",
+        schedule_timeout: str = "5m",
+        sticky: StickyStrategy | None = None,
+        retries: int = 0,
+        rate_limits: list[RateLimit] = [],
+        desired_worker_labels: dict[str, DesiredWorkerLabel] = {},
+        concurrency: ConcurrencyExpression | None = None,
+        on_failure: Union["Function[R]", None] = None,
+        default_priority: int = 1,
+        input_validator: Type[TWorkflowInput] | None = None,
+        backoff_factor: float | None = None,
+        backoff_max_seconds: int | None = None,
```

Comment:
one question here is if we want to prefix these or something, since some pertain to the workflow, some pertain to the step, and some to both (IMO this is confusing)

Reply:
i think one of the goals we should consider is to make everything effectively run as a step so all features are always in scope. in other words, a dag is a step that calls steps (somehow...). ultimately, i think the hatchet service should still be responsible for dags but i think we can orchestrate through the main scheduler. i know @abelanger5 has thoughts here.

we might want to feature space some of these things to simplify?

i.e.:
```
on: TriggerConfig
retries: int | RetryConfig
```

Reply:
I have mixed feelings here I think. I sorta like the class-based thing for defining DAGs like we have it now, because it's pretty clear what's related to what, and I'm nervous that without that we'd just end up with a mishmash of random functions hanging around that all depend on each other, which could be hard to untangle. That said, I do really like being able to just do `@function` or whatever and have it all kinda magically work.

I'm sure there's a way to make this work nicely though (I don't think there's any reason that e.g. you couldn't specify `parents` on a `function` or something, similar to how we have it now). Happy to discuss some more here!

re: the second thing - feature spacing sounds great to me, I think this makes a lot of sense 👍